src/StreamJsonRpc/JsonMessageFormatter.cs

```diff
@@ -518,5 +649,106 @@ internal JsonArrayPool(ArrayPool<T> arrayPool)
 
             public void Return(T[] array) => this.arrayPool.Return(array);
         }
+
+        private class JsonProgressClientConverter : JsonConverter
+        {
+            private readonly JsonMessageFormatter formatter;
+
+            public JsonProgressClientConverter(JsonMessageFormatter formatter)
+            {
+                this.formatter = formatter ?? throw new ArgumentNullException(nameof(formatter));
+            }
+
+            public override bool CanConvert(Type objectType) => MessageFormatterHelper.FindIProgressOfT(objectType) != null;
+
+            public override object ReadJson(JsonReader reader, Type objectType, object existingValue, JsonSerializer serializer)
+            {
+                throw new NotSupportedException();
+            }
+
+            public override void WriteJson(JsonWriter writer, object value, JsonSerializer serializer)
+            {
+                // if the requestId is empty it means the Progress object comes from a response or a notification
+                if (this.formatter.requestId == null)
+                {
+                    throw new NotSupportedException("IProgress<T> objects should not be part of any response or notification.");
+                }
+
+                lock (this.formatter.progressMap)
+                {
+                    long progressId = this.formatter.nextProgressId++;
+                    this.formatter.requestProgressMap.Add(this.formatter.requestId.Value, progressId);
+
+                    this.formatter.progressMap.Add(progressId, new ProgressParamInformation(value));
+
+                    writer.WriteValue(progressId);
+                }
+            }
+        }
+
+        private class ProgressParamInformation
+        {
+            public Type iProgressOfTType;
```

Comment:
>public [](start = 12, length = 6)

nit: avoid using `public` for members of non-public types, since they're not actually public it's misleading. Can you use `internal` for all of these instead?

Reply:
Also: they should be `readonly` I think, right? And StyleCop won't like that you have fields that aren't private, so can you make them readonly properties instead? (e.g. `internal Type IProgressOfTType { get; }`)

---
In reply to: [304481872](https://github.com/milopezc/vs-streamjsonrpc/pull/1#discussion_r304481872) [](ancestors = 304481872)